compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/EncodeV1.hs

```diff
@@ -39,26 +42,42 @@ newtype WithInterning = WithInterning{getWithInterning :: Bool}
 data EncodeEnv = EncodeEnv
     { version :: !Version
     , withInterning :: !WithInterning
-    , internedStrings :: !(HMS.HashMap T.Text Word64)
+    , internedStrings :: !(HMS.HashMap T.Text Int32)
+    , internedDottedNames :: !(HMS.HashMap [Int32] Int32)
     }
 
 initEncodeEnv :: Version -> WithInterning -> EncodeEnv
-initEncodeEnv version withInterning = EncodeEnv{internedStrings = HMS.empty, ..}
+initEncodeEnv version withInterning =
+    EncodeEnv{internedStrings = HMS.empty, internedDottedNames = HMS.empty, ..}
 
 -- | Find or allocate a string in the interning table. Return the index of
 -- the string in the resulting interning table.
-allocString :: T.Text -> Encode Word64
+allocString :: T.Text -> Encode Int32
 allocString t = do
     env@EncodeEnv{internedStrings} <- get
     case t `HMS.lookup` internedStrings of
         Just n -> pure n
         Nothing -> do
-            -- NOTE(MH): We assumie that the number of interned strings fits
-            -- in a `Word64`. (More than that would require A LOT of memory.)
-            let n = fromIntegral (HMS.size internedStrings)
+            let m = HMS.size internedStrings
+            when (m > fromIntegral (maxBound :: Int32)) $
+                error "String interning table grew too large"
+            let n = fromIntegral m :: Int32
             put $! env{internedStrings = HMS.insert t n internedStrings}
             pure n
 
+allocList :: [Int32] -> Encode Int32
+allocList ids = do
+    env@EncodeEnv{internedDottedNames} <- get
+    case ids `HMS.lookup` internedDottedNames of
+        Just n -> pure n
+        Nothing -> do
+            let m = HMS.size internedDottedNames
```

Comment:
`HMS.size` is `O(n)` so this is quadratic in the number of strings you intern. You probably want to keep track of the next free id in `EncodeEnv` to avoid this (there are other places that suffer from the same problem), e.g., `allocString`.

Reply:
I've added fields `nextInternedStringId` and `nextInternedDottedNameId` to the `EncodeEnv` to track these numbers explicitly.